### ISSUE TITLE
Need for assumptions

### ISSUE THREAD

**Issue by fedeinthemix** [JuliaSymbolics/SymbolicUtils.jl#228]
If this library is intended to become the basis for a full CAS I think it should include a facility to define some form of assumptions. 

Consider the simple example of solving the linear equation `a * x = b`, a, b, x Real. The implicit assumption in Symbolics is that things will go right
```julia
@variables x a b
Symbolics.solve_for([a*x ~ b], [x])
```
gives `b*(a^-1)`. However, if a and b are both zero then the equation is satisfied for arbitrary x. Conversely, if a is zero, and b != 0 the there is no solution. Just plugging values for b and a in the result is not a solution as, for example, `0*(0^-1)` and `1*(0^-1)` give NaN and Inf respectively which are hardly interpreted as arbitrary value and no value.

Another trivial example is integration of `1/x^n`, n an Integer. Need to distinguish n=1.

Given that you need to distinguish some particular condition or value of a variable of a given type, it likely can't be handled by the type system.

**Comment by ChrisRackauckas** [comment]
Agreed. @shashi we should probably discuss the form of this. I think it might be good to use DomainSets.jl for formalizing the domain specifications.

**Comment by shashi** [comment]
Also see [SymbolicSAT.jl](https://github.com/JuliaSymbolics/SymbolicSAT.jl) needs some resurrecting.
